br/pkg/restore/snap_client/pipeline_items.go

```diff
@@ -118,154 +116,262 @@ func (rc *SnapClient) RestorePipeline(ctx context.Context, plCtx PipelineContext
-		err = multierr.Append(err, multierr.Combine(ExhaustErrors(errCh)...))
-	case <-finish:
+	// the middle pipeline tasks
+	for _, f := range builder.pipelineFunctions {
+		postHandleCh = handler.concurrentHandleTablesCh(postHandleCh, f.concurrency, f.taskLabel, f.processFn, f.deferFn)
 	}
 
-	return errors.Trace(err)
+	// the last pipeline task
+	handler.dropToBlackhole(postHandleCh)
+
+	return eg.Wait()
+}
+
+type PipelineConcurrentHandler struct {
+	pipelineTaskCtx context.Context
+	eg              *errgroup.Group
 }
 
-func afterTableRestoredCh(ctx context.Context, createdTables []*CreatedTable) <-chan *CreatedTable {
+func (handler *PipelineConcurrentHandler) afterTableRestoredCh(createdTables []*CreatedTable) <-chan *CreatedTable {
 	outCh := make(chan *CreatedTable)
 
-	go func() {
+	handler.eg.Go(func() error {
 		defer close(outCh)
 
 		for _, createdTable := range createdTables {
 			select {
-			case <-ctx.Done():
-				return
-			default:
-				outCh <- createdTable
+			case <-handler.pipelineTaskCtx.Done():
+				return handler.pipelineTaskCtx.Err()
+			case outCh <- createdTable:
 			}
 		}
-	}()
+
+		return nil
+	})
 	return outCh
 }
 
 // dropToBlackhole drop all incoming tables into black hole,
 // i.e. don't execute checksum, just increase the process anyhow.
-func dropToBlackhole(ctx context.Context, inCh <-chan *CreatedTable, errCh chan<- error) <-chan struct{} {
-	outCh := make(chan struct{}, 1)
-	go func() {
-		defer func() {
-			close(outCh)
-		}()
+func (handler *PipelineConcurrentHandler) dropToBlackhole(inCh <-chan *CreatedTable) {
+	handler.eg.Go(func() error {
 		for {
 			select {
-			case <-ctx.Done():
-				errCh <- ctx.Err()
-				return
+			case <-handler.pipelineTaskCtx.Done():
+				return handler.pipelineTaskCtx.Err()
 			case _, ok := <-inCh:
 				if !ok {
-					return
+					return nil
 				}
 			}
 		}
-	}()
-	return outCh
+	})
 }
 
-func concurrentHandleTablesCh(
-	ctx context.Context,
+func (handler *PipelineConcurrentHandler) concurrentHandleTablesCh(
 	inCh <-chan *CreatedTable,
-	outCh chan<- *CreatedTable,
-	errCh chan<- error,
-	workers *tidbutil.WorkerPool,
+	concurrency uint,
+	taskLabel string,
 	processFun func(context.Context, *CreatedTable) error,
-	deferFun func()) {
-	eg, ectx := errgroup.WithContext(ctx)
-	defer func() {
-		if err := eg.Wait(); err != nil {
-			errCh <- err
-		}
-		close(outCh)
-		deferFun()
-	}()
-
-	for {
-		select {
-		// if we use ectx here, maybe canceled will mask real error.
-		case <-ctx.Done():
-			errCh <- ctx.Err()
-		case tbl, ok := <-inCh:
-			if !ok {
+	deferFun func(context.Context) error,
+) (outCh chan *CreatedTable) {
+	outCh = defaultOutputTableChan()
+	handler.eg.Go(func() (pipelineErr error) {
+		workers := tidbutil.NewWorkerPool(concurrency, taskLabel)
+		eg, ectx := errgroup.WithContext(handler.pipelineTaskCtx)
+		defer func() {
+			// Note: directly return the error and then the pipelineTaskCtx will be cancelled.
+			if err := eg.Wait(); err != nil {
```

Comment:
should it execute next deferFun when met error here?

Reply:
We can regard the process as
```
processFun
processFun
...
processFun       <- error occurs
processFun       <- cancel
processFun       <- cancel
...              <- eg.Wait()
processFun       <- skip
processFun       <- skip
...
deferFun         <- should also skip
```